[Android] Entry & Picker VerticalTextAlignment ignored

kubaflo opened the pull request:
### Issues Fixed

Fixes https://github.com/dotnet/maui/issues/21717

|Before|After|
|--|--|
|<img src="https://github.com/dotnet/maui/assets/42434498/39a2fc72-7ff8-45bb-90b3-587cc77034e6" width="300px"/>|<img src="https://github.com/dotnet/maui/assets/42434498/da538d06-d711-4c47-bcce-3a26ef95f091" width="300px"/>|

jsuarezruiz commented:
/azp run

jsuarezruiz commented on src/Controls/tests/UITests/Tests/Issues/Issue21717.cs:
```diff
@@ -0,0 +1,21 @@
+		{
+			this.IgnoreIfPlatforms(new[] { TestDevice.Mac, TestDevice.iOS, TestDevice.Windows });
+
+			VerifyScreenshot();
```
Could you assign an AutomationId to the Picker and wait for it before take and verify the snapshot?

![image](https://github.com/dotnet/maui/assets/6755973/9f8e3176-86ad-4c93-8e8e-9acd4ec1bed5)

kubaflo replied:
Yes, my bad. I added a commit with it

jsuarezruiz commented:
/azp run

mattleibow commented on src/Core/src/Handlers/Picker/PickerHandler.Android.cs:
```diff
@@ -165,5 +167,14 @@ static void Reload(IPickerHandler handler)
+		{
+			if (PlatformView != null && VirtualView != null)
+			{
+				PlatformView.SetHeight(PlatformView.Height);
```
Not sure I understand what this line is doing?

kubaflo replied:
Yes, this is a strange one, but it fixes it. For some reason, these heights have different values

mattleibow replied:
What are the different values? Is one really incorrect but still a valid size? Like before it is 30 but the correct height is 300, or is it 0 and 300?

kubaflo replied:
It is 0 and then 300

mattleibow commented on src/Core/src/Handlers/Picker/PickerHandler.Android.cs:
```diff
@@ -165,5 +167,14 @@ static void Reload(IPickerHandler handler)
+			if (PlatformView != null && VirtualView != null)
+			{
+				PlatformView.SetHeight(PlatformView.Height);
+				PlatformView.UpdateVerticalTextAlignment(VirtualView);
```
Instead of calling the Update directly, can you rather invoke the mapper? This way if someone overrides the mapper, their code still gets called.

mattleibow replied:
```suggestion
				Handler.UpdateValue(nameof(IEntry.VerticalTextAlignment));
```

jsuarezruiz commented:
/azp run

jsuarezruiz commented:
/azp run

jsuarezruiz commented:
Updated test snapshot.

jsuarezruiz commented:
/azp run

jsuarezruiz commented on src/Core/src/Handlers/Entry/EntryHandler.Android.cs:
```diff
@@ -274,5 +276,14 @@ internal void HideClearButton()
+		{
+			if (PlatformView != null && VirtualView != null)
+			{
+				PlatformView.SetHeight(PlatformView.Height);
```
Maybe related with this change, some Entry tests are failing
![image](https://github.com/dotnet/maui/assets/6755973/da535d4e-7cf8-41ae-a5d7-c44ece1f3e66)

kubaflo replied:
Hmm strange, they pass for me, and the last one is about verifying screenshots which also looks good to me
<img width="570" alt="Screenshot 2024-06-15 at 01 51 04" src="https://github.com/dotnet/maui/assets/42434498/83de3439-f533-4a45-b60f-a3e255fdb7b3">
![Screenshot_1718409338](https://github.com/dotnet/maui/assets/42434498/17d67dc7-9070-45f9-adab-d1339e09607d)

jsuarezruiz commented:
/azp run

mattleibow commented:
@kubaflo did you see this comment? https://github.com/dotnet/maui/issues/21717#issuecomment-2333824320

kubaflo commented:
> @kubaflo did you see this comment? [#21717 (comment)](https://github.com/dotnet/maui/issues/21717#issuecomment-2333824320)
> 
> > I tookover the planned fix for #21966 into my EntryHandler (https://github.com/dotnet/maui/pull/21966/files#diff-04130ba6aadaa5a68624b127e29d3f85384872fc0520ed02821a5b96ac78ecbb) and my result is:
> > 
> > * Working for API 31
> > * Not working for API 34, if within ScrollView, and Entry is created in code behind and properties set in code behind before adding to a parent.
> > * Also note LayoutChange is called while Scrolling - performance!
> > * Note we have nested ScrollViews - (I know it is not good UI practise but it should work for this case also)
> > 
> > This here works for me - but is still not good:
> > ```
> >    private bool m_IsInitialized;
> >    private int? m_Height;
> >    private void PlatformView_LayoutChange(object sender, Android.Views.View.LayoutChangeEventArgs e)
> >    {
> >        // only do this once (otherwise side effects e. g. when scrolling)
> >        PlatformView.LayoutChange -= PlatformView_LayoutChange;
> > 
> >        // .NET Maui Workaround bug because VerticalTextAlignment is not working on Android within Grid
> >        if ((PlatformView != null) && VirtualView != null)
> >        {
> >            if (m_Height == PlatformView.Height)
> >            {
> >                return;
> >            }
> > 
> >            m_Height = PlatformView.Height;
> > 
> >            // for API 31
> >            PlatformView.SetHeight(PlatformView.Height);
> > 
> >            if (!m_IsInitialized)
> >            {
> >                // for API 34
> >                m_IsInitialized = true;
> >                ((UnifiedHmiEntry)VirtualView).Dispatcher.Dispatch(() =>
> >                {
> >                    try
> >                    {
> >                        PlatformView.SetHeight(PlatformView.Height);
> >                    }
> >                    catch (Exception exception)
> >                    {
> >                    }
> >                });
> >            }
> >        }
> >    }
> > ```

Hmm interesting... I will investigate it

mattleibow commented on src/Core/src/Handlers/Picker/PickerHandler.Android.cs:
```diff
@@ -29,6 +30,7 @@ protected override void DisconnectHandler(MauiPicker platformView)
 		{
 			platformView.FocusChange -= OnFocusChange;
 			platformView.Click -= OnClick;
+			platformView.LayoutChange -= PlatformViewLayoutChange;
```
This will fire for everything as noted by the comment https://github.com/dotnet/maui/issues/21717#issuecomment-2333824320

Does this need to be all the time or just the first time? What happens if you changed the grid row height from say 200 to 400, is the alignment wrong again?

mattleibow reviewed (changes requested):
I added some comments and questions again, but this is really worrying as the layout should be setting these values.

Unless, the picker is in a wrapper view and somehow we are not doing layout correctly in that case. But I don't think that is the case since there would be no need for a wrapper.

Not sure the best way to do this...

kubaflo committed:
[Android] Entry & Picker VerticalTextAlignment ignored

kubaflo commented:
@mattleibow Look what I've found
https://github.dev/kubaflo/maui/blob/9756ef954e96a1e0325e05e028b1639a2ffe4fca/src/Core/src/Handlers/ViewHandlerExtensions.Android.cs#L146-L172

This is currently used for the editor, and I've applied it to the picker and entry and it fixes the issue

kubaflo committed:
PR #21966 Review: Approve - Fix properly hooks into PlatformArrange lifecycle

- Verified tests correctly reproduce bug (Phase 2-3)
- Explored independent fix alternative via try-fix (Phase 4)
- Alternative (RequestLayout) failed due to timing - needs actual layout completion
- PR's approach is optimal: hooks PlatformArrange (correct lifecycle point)
- Reuses existing Editor pattern (PrepareForTextViewArrange)
- Fixed test: Removed duplicate Category attribute (analyzer violation)

Final Recommendation: APPROVE

kubaflo committed:
Update Issue21717.cs

StephaneDelcroix reviewed (commented):
## Multi-Model Consensus Review (5 models: claude-opus-4.6 ×2, claude-sonnet-4.6, gemini-3-pro-preview, gpt-5.3-codex)

### CI Status: ⚠️ Pre-existing (Build Analysis meta-job fails; all actual build/test jobs pass ✅)

### Consensus Findings
No consensus bugs found. All 5 models agreed the fix is correct.

The fix correctly overrides `PlatformArrange` in both `EntryHandler` and `PickerHandler` on Android to call `PrepareForTextViewArrange(frame)` before delegating to base -- the same established pattern already used by `EditorHandler` in `ViewHandlerExtensions.Android.cs`.

### Test Coverage
✅ Full UI test added (`Issue21717`) with screenshot baseline for Android device profile.

### Minor Note (1 model only -- did not reach consensus)
`Task.Delay(500)` before `VerifyScreenshot` could use `retryTimeout` parameter instead -- style preference, not a bug.

### Verdict: ✅ Approve
Correct, minimal fix using an established pattern. Gate-verified (tests fail without fix, pass with fix).

StephaneDelcroix reviewed (approved):
fix the conflict, make sure it's green
